### PR TITLE
Deprecate smartBinarySearch

### PR DESCRIPTION
**Pull request by data-man**
- binarySearch improved
- smartBinarySearch marked as deprecated

### CODE REVIEW

**Comment by Araq** [comment]
Good idea, but is this based on benchmarks? maybe this "smart" version is not smart in practice?

**Comment by data-man** [comment]
```nim
import nimbench, os, algorithm, strutils, random

var
  a: seq[int]
  length: int = 10
  needle: int = 10

if paramCount() > 0:
  length = paramStr(1).parseInt()

  if paramCount() > 1:
    needle = paramStr(2).parseInt()

a = newSeq[int](length)

for i in 0..< a.len:

  a[i] = rand(a.len)

sort(a, cmp)

echo length
echo needle
#echo a

bench(binarySearch):
  var
    pos: int

  pos = binarySearch(a, needle)

  doNotOptimizeAway(pos)

benchRelative(smartBinarySearch):
  var
    pos: int

  pos = smartBinarySearch(a, needle)

  doNotOptimizeAway(pos)

runBenchmarks()
```
with 1000000 13 parameters

GlobalBenchmark |                                relative | time/iter | iters/s
------------ | ------------- | ------- | -----
GlobalBenchmark |                               ||              5.22ns | 191.55M
bench.nim    |                                   relative | time/iter | iters/s
binarySearch          |                                   |  423.08ns  |  2.36M
smartBinarySearch                 |               775.97%  |  54.52ns |  18.34M

**Comment by skilchen** [comment]
please repeat the benchmark using -d:release

File: lib/pure/algorithm.nim [hunk 1]
```diff
@@ -66,18 +66,32 @@ proc reversed*[T](a: openArray[T]): seq[T] =
...
+    while step > 0:
+      let i = result or step
+      if likely(a[i] <= key):
+        result = i
+      step = step shr 1
+    if a[result] != key: result = -1
+  else:
+    while result < b:
+      var mid = (result + b) shr 1
+      if likely(a[mid] < key):
```
Comment (Araq) [inline lib/pure/algorithm.nim]: Why would this be "likely" for unknown data sets? Seems like a 50-50 chance.

**Comment by Araq** [comment]
A difference of 2.3M vs 18.34M needs to be explained. If true, this means people should be aware to use powers of two for the arrays they use for binary search. And then it makes more sense to keep the specialized algorithm around.

**Comment by dom96** [comment]
Wouldn't it make sense for ``smartBinarySearch`` to assert that the length of the array is a power of two?

**Comment by Araq** [comment]
> Wouldn't it make sense for smartBinarySearch to assert that the length of the array is a power of two?

Certainly.

**Comment by Parashurama** [comment]
I use this code when I need Binary search on array pow of 2 array.
It's much faster than regular binary search (no branch mis-prediction)
There is also an adaptation for non pow of 2, but it require fastlog2 / CLZ (CTZ) to be efficient.
see blog post: https://blog.demofox.org/2017/06/20/simd-gpu-friendly-branchless-binary-search/
```nim
# warning: this is translated code from working C code so this need to be verified.
proc smartBinarySearch*[T](a: openArray[T], key: T): int =
  var ret = 0
  var hlen= a.len div 2
  while hlen>16:
    ret += (a[ret + hlen] <= key) ? hlen : 0;
    hlen = hlen div 2
  
  ret += (a[ret + 16] <= key).ord * 16;
  ret += (a[ret +  8] <= key).ord * 8;
  ret += (a[ret +  4] <= key).ord * 4;
  ret += (a[ret +  2] <= key).ord * 2;
  ret += (a[ret +  1] <= key).ord * 1;
  return if a[ret] == key: ret else: -1;
```

**Comment by data-man** [comment]
My idea is to make the binarySearch more universal and adapted to specific cases.
Suggestions:
- deprecate smartBinarySearch. I doubt that someone is using it. :)
- add cmp parameter to binarySearch, as in the sort.
E.g. there is base seq[T], where T is tuple or object.
2st tuple sorted by field1, 3rd - by field2.
Then the binarySearch (or smartBinarySearch) is useless for searching.

**Comment by konqoro** [comment]
I was going to ask if the benchmark was actually returning the correct result (but I assume it was checked). I agree with @data-man there is a lot of duplication of binarySearch in the stdlib and something could be done about it.

**Comment by Araq** [comment]
> Secondly, I agree with @data-man there is a lot of duplication of binarySearch in the stdlib and something could be done about it.

It exists twice in the entire stdlib or something.

**Comment by konqoro** [comment]
ok in colors.nim, sysstr.nim and unicode.nim

**Comment by data-man** [comment]
In compiler - binaryStrSearch.

File: lib/pure/algorithm.nim [hunk 2]
```diff
@@ -66,18 +66,32 @@ proc reversed*[T](a: openArray[T]): seq[T] =
...
   var b = len(a)
-  while result < b:
-    var mid = (result + b) div 2
-    if a[mid] < key: result = mid + 1
-    else: b = mid
-  if result >= len(a) or a[result] != key: result = -1
-
-proc smartBinarySearch*[T](a: openArray[T], key: T): int =
+  if b == 0: return -1
+  if unlikely((b and (b - 1)) == 0):
```
Comment (krux02) [inline lib/pure/algorithm.nim]: Why is this unlikely? I know a lot of data is constructed to have a power of two size which makes this less "unlikely". I would leave it as the default.

**Comment by krux02** [comment]
You forgot two things: Deprecated stuff should always point the alternatives. Changelog entry.

**Comment by krux02** [comment]
And for the discussion, we should all agree that "smartBinarySearch" is a stupid name. Nothing should be calld "smart...", "secure...", "correct..." or any of these other prefixes. They are all wrong. Just recently "secure"-prefixed functions have been deprecated.

**Pull request closed without merging**